File: superleague-ruby/superleague.rb
#!/usr/bin/env ruby

require "json"

class Table
  def initialize(matches)
    rows = []
    matches.each do |m|
      hg = m["homeGoals"].to_i
      ag = m["awayGoals"].to_i
      rows << TableRow.new(m["homeTeam"], hg, ag)
      rows << TableRow.new(m["awayTeam"], ag, hg)
    end
    table = {}
    rows.each do |r|
      if table.key?(r.team)
        table[r.team].merge(r)
      else
        table[r.team] = r
      end
    end
    rows = table.values
    rows.sort_by! { |r| [-r.points, -r.goals_difference, -r.wins, r.team] }
    for i in 1..rows.length
      rows[i-1].rank = i
    end
    @table_rows = rows
  end
  def format()
    lines = []
    title_fmt = "%25s %2s %2s %2s %2s %2s %2s %3s %2s"
    row_fmt = "%25s %2d %2d %2d %2d %2d %2d %3d %2d"
    title_line = title_fmt % ["Team", "#", "w", "l", "d", "+", "-", "=", "P"]
    lines << title_line
    lines << "-" * title_line.length
    @table_rows.each do |r|
      lines << row_fmt % [r.team, r.rank, r.wins, r.defeats, r.ties,
                          r.goals_scored, r.goals_conceded, r.goals_difference,
                          r.points]
    end
    return lines.join("\n")
  end
end

class TableRow
  attr_accessor :team, :rank, :wins, :defeats, :ties,
    :goals_scored, :goals_conceded, :goals_difference, :points
  def initialize(team, goals_scored, goals_conceded)
    @team = team
    @rank = -1
    @goals_scored = goals_scored
    @goals_conceded = goals_conceded
    @goals_difference = goals_scored - goals_conceded
    if goals_scored > goals_conceded
      @wins = 1
      @defeats = 0
      @ties = 0
      @points = 3
    elsif goals_scored < goals_conceded
      @wins = 0
      @defeats = 1
      @ties = 0
      @points = 0
    else
      @wins = 0
      @defeats = 0
      @ties = 1
      @points = 1
    end
  end
  def merge(other)
    @goals_scored += other.goals_scored
    @goals_conceded += other.goals_conceded
    @goals_difference += other.goals_difference
    @wins += other.wins
    @defeats += other.defeats
    @ties += other.ties
    @points += other.points
  end
end

matches_file = File.open(ARGV[0])
matches_data = matches_file.read
matches = JSON.parse(matches_data)
table = Table.new(matches)
puts table.format()

File: superleague-python/superleague.py
#!/usr/bin/env python3

import json
import sys


def create_table(json_file_path):
    with open(sys.argv[1], 'r') as f:
        data = f.read()
        matches = json.loads(data)

    rows = {}

    for result in matches:

        homeTeam = result['homeTeam']
        awayTeam = result['awayTeam']
        homeGoals = result['homeGoals']
        awayGoals = result['awayGoals']

        if homeTeam not in rows:
            rows[homeTeam] = new_row(homeTeam)
        if awayTeam not in rows:
            rows[awayTeam] = new_row(awayTeam)

        rows[homeTeam]['goalsScored'] += homeGoals
        rows[homeTeam]['goalsConceded'] += awayGoals
        rows[awayTeam]['goalsScored'] += awayGoals
        rows[awayTeam]['goalsConceded'] += homeGoals

        if homeGoals > awayGoals:
            rows[homeTeam]['wins'] += 1
            rows[awayTeam]['defeats'] += 1
        elif homeGoals < awayGoals:
            rows[awayTeam]['wins'] += 1
            rows[homeTeam]['defeats'] += 1
        else:
            rows[homeTeam]['ties'] += 1
            rows[awayTeam]['ties'] += 1

    for row in rows.values():
        row['goalsDifference'] = row['goalsScored'] - row['goalsConceded']
        row['points'] = row['wins'] * 3 + row['ties']

    rows = rows.values()
    rows = sorted(rows, key=lambda r: r['team'])
    rows = sorted(rows, key=lambda r: r['wins'], reverse=True)
    rows = sorted(rows, key=lambda r: r['goalsDifference'], reverse=True)
    rows = sorted(rows, key=lambda r: r['points'], reverse=True)

    for i, row in enumerate(rows):
        row['rank'] = i + 1

    return rows


def new_row(team):
    return {
        'team': team,
        'rank': 0,
        'wins': 0,
        'ties': 0,
        'defeats': 0,
        'goalsScored': 0,
        'goalsConceded': 0,
        'goalsDifference': 0,
        'points': 0,
    }

def output_table(rows):
    print('                     Name  #  w  d  l  +  -   =  P')
    print('--------------------------------------------------')
    for row in rows:
        print(f'{row["team"]:>25s} {row["rank"]:>2d} ' +\
              f'{row["wins"]:>2d} {row["ties"]:>2d} {row["defeats"]:>2d} ' +\
              f'{row["goalsScored"]:>2d} {row["goalsConceded"]:>2d} {row["goalsDifference"]:>3d} ' +\
              f'{row["points"]:>2d}')


if __name__ == '__main__':
    table = create_table(sys.argv[1])
    output_table(table)
